router/protocol.go

```diff
@@ -2,58 +2,329 @@ package router
+	// Write in a separate goroutine to avoid the possibility of
+	// deadlock
+	sendHeader := append(ProtocolBytes, params.MinVersion, params.MaxVersion)
+	writeDone := make(chan error, 1)
+	go func() {
+		_, err := params.Conn.Write(sendHeader)
+		writeDone <- err
+	}()
+
+	header := make([]byte, len(ProtocolBytes)+2)
+	if n, err := io.ReadFull(params.Conn, header); err != nil && n == 0 {
 		return 0, fmt.Errorf("failed to receive remote protocol header: %s", err)
 	} else if err != nil {
 		return 0, fmt.Errorf("received incomplete remote protocol header (%d octets instead of %d): %v; error: %s",
-			n, ProtocolHeaderLen, header[:n], err)
+			n, len(header), header[:n], err)
 	}
-	if !bytes.Equal(ProtocolBytes, header[:ProtocolLen]) {
-		return 0, fmt.Errorf("remote protocol header not recognised: %v", header[:ProtocolHeaderLen])
+
+	if !bytes.Equal(ProtocolBytes, header[:len(ProtocolBytes)]) {
+		return 0, fmt.Errorf("remote protocol header not recognised: %v", header[:len(ProtocolBytes)])
 	}
-	var (
-		remoteMinVersion = header[ProtocolLen]
-		remoteVersion    = header[ProtocolLen+1]
-	)
-	minVersion := ProtocolMinVersion
-	if remoteMinVersion > minVersion {
-		minVersion = remoteMinVersion
+
+	theirMinVersion := header[len(ProtocolBytes)]
+	minVersion := theirMinVersion
+	if params.MinVersion > minVersion {
+		minVersion = params.MinVersion
 	}
-	maxVersion := ProtocolVersion
-	if remoteVersion < maxVersion {
-		maxVersion = remoteVersion
+
+	theirMaxVersion := header[len(ProtocolBytes)+1]
+	maxVersion := theirMaxVersion
+	if maxVersion > params.MaxVersion {
+		maxVersion = params.MaxVersion
 	}
+
 	if minVersion > maxVersion {
-		return 0, fmt.Errorf("remote version [%d,%d] is incompatible with our version [%d,%d]",
-			remoteMinVersion, remoteVersion, ProtocolMinVersion, ProtocolVersion)
+		return 0, fmt.Errorf("remote version range [%d,%d] is incompatible with ours [%d,%d]",
+			theirMinVersion, theirMaxVersion,
+			params.MinVersion, params.MaxVersion)
+	}
+
+	if err := <-writeDone; err != nil {
+		return 0, err
 	}
+
 	return maxVersion, nil
 }
 
+// The V1 procotol consists of the protocol identification/version
+// header, followed by a stream of gobified values.  The first value
+// is the encoded features map (never encrypted).  The subsequent
+// values are the messages on the connection (encrypted for an
+// encrypted connection).  For an encrypted connection, the public key
+// is passed in the "PublicKey" feature as a string of hex digits.
+func (res *ProtocolIntroResults) doIntroV1(params ProtocolIntroParams, pubKey, privKey *[32]byte) error {
+	features := filterV1Features(params.Features)
+	if pubKey != nil {
+		features["PublicKey"] = hex.EncodeToString(pubKey[:])
+	}
+
+	enc := gob.NewEncoder(params.Conn)
+	dec := gob.NewDecoder(params.Conn)
+
+	// Encode in a separate goroutine to avoid the possibility of
+	// deadlock
+	encodeDone := make(chan error, 1)
```

Comment:
Is this 1-capacity in order to prevent the goroutine from getting stuck and hanging around forever in the event of an error in `dec.Decode` below? If so, that merits a comment. If not, then make it 0.